vendor/lib/src/plan.dart

```diff
@@ -0,0 +1,142 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:vendor/src/config.dart' show VendorState, VendorConfig;
+import 'package:vendor/src/action/action.dart' show Action;
+import 'package:vendor/src/utils/map_ext.dart' show MapExt;
+
+/// Plan the actions required to go from [currentState] to [desiredState].
+List<Action> plan({
+  required String rootPackageName,
+  required VendorConfig desiredState,
+  required VendorConfig currentState,
+  required Uri defaultHostedUrl,
+}) {
+  bool notSameSameSource(String name) {
+    final current = currentState.dependencies[name];
+    final desired = desiredState.dependencies[name];
+    return current != desired &&
+        ((current == null || desired == null) ||
+            current.package != desired.package ||
+            current.version != desired.version ||
+            (!current.include.containsAll(desired.include) ||
+                !desired.include.containsAll(current.include)));
+  }
+
+  return [
+    // Vendored folders to be removed
+    ...currentState.dependencies.keys
+        .where(notSameSameSource)
+        .map((name) => Action.removeFolder(_vendoredPackageFolder(name))),
```

Comment:
Not sure about this case. Do we also need to do rewrites at the top-level here?

Reply:
Not really,

It's possible that I just changed the version of a vendored package. Then we have to remove the folder and re-create it later.

The top-level rewrites are handle separately... actually, I don't think the `plan` function cares if the thing you rewrite to even exists. I think there is something in `validate.dart` that checks that the rewrites in the desired configuration actually makes sense.

Reply:
I think I had the assumption that a vendored package implied an implicit top-level rewrite. But I guess it is not so.